### PR TITLE
reflector : remove trait AxVMHal

### PR DESCRIPTION
**Pull request by pengzechen**
移除：
arceos-vmm/src/hal.rs
crates/axvm/src/hal.rs
添加：
crates/axvmhal
axvmhal中定义了与arceos相关trait并实现了这些trait，解决了AxVMhal 泛型扩散的问题。
arceos-vmm 引用axvmhal，axvm 引用axvmhal
注意：目前只解决了aarch64 的问题

### CODE REVIEW

File: crates/axvm/src/mm/mod.rs [hunk 1]
```diff
@@ -28,20 +27,18 @@ pub struct NestedPageFaultInfo {
 /// A 4K-sized contiguous physical memory page, it will deallocate the page
 /// automatically on drop.
 #[derive(Debug)]
-pub struct PhysFrame<H: AxVMHal> {
+pub struct PhysFrame {
```
Comment (hky1999) [inline crates/axvm/src/mm/mod.rs:30]: The `PhysFrame` will be removed soon.

File: crates/axvm/src/vcpu.rs [hunk 2]
```diff
@@ -234,25 +232,25 @@ impl<H: AxVMHal> AxVCpu<H> {
 #[percpu::def_percpu]
 static mut CURRENT_VCPU: Option<*mut u8> = None;
 
-pub fn get_current_vcpu<H: AxVMHal>() -> Option<&'static AxVCpu<H>> {
+pub fn get_current_vcpu() -> Option<&'static AxVCpu> {
```
Comment (hky1999) [inline crates/axvm/src/vcpu.rs:235]: `CURRENT_VCPU` will be removed soon.

File: crates/axvm/src/vm.rs [hunk 3]
```diff
@@ -56,7 +57,7 @@ impl<H: AxVMHal> AxVM<H> {
                     vcpu_id,
                     pcpu_id,
                     pcpu_id,
-                    <AxArchVCpuImpl<H> as AxArchVCpu>::CreateConfig::default(),
+                    <AxArchVCpuImpl as AxArchVCpu>::CreateConfig::default(),
```
Comment (hky1999) [inline crates/axvm/src/vm.rs:60]: `CreateConfig` and `SetupConfig` will be removed soon.

File: crates/axvm/src/vm.rs [hunk 4]
```diff
@@ -125,7 +125,7 @@ impl<H: AxVMHal> AxVM<H> {
             vcpu.setup(
                 entry.as_usize(),
                 result.ept_root(),
-                <AxArchVCpuImpl<H> as AxArchVCpu>::SetupConfig::default(),
+                <AxArchVCpuImpl as AxArchVCpu>::SetupConfig::default(),
```
Comment (hky1999) [inline crates/axvm/src/vm.rs:128]: CreateConfig and SetupConfig will be removed soon.

File: crates/axvmhal/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+pub type GuestPhysAddr = usize;
+/// Host virtual address.
+pub type HostVirtAddr = VirtAddr;
+/// Host physical address.
+pub type HostPhysAddr = PhysAddr;
+pub use memory_addr::PAGE_SIZE_4K as PAGE_SIZE;
+
+/// The interfaces which the underlying software (kernel or hypervisor) must implement.
+
+ pub trait AxVMHal: Sized + Clone {
```
Comment (hky1999) [inline crates/axvmhal/src/lib.rs:18]: After a discussion with @equation314 (he suggested  that [crate_interface](https://crates.io/crates/crate_interface) can be used here). 

Placing all these interfaces into a single trait is not a good design. 
It's better to use a super trait or an associated type， @aarkegz 

e.g.
`PagingHandler` for [axaddrspace](https://github.com/arceos-hypervisor/axaddrspace)
`VMMHandler` for semantics like `get_vm_by_id()`

Comment (aarkegz) [reply crates/axvmhal/src/lib.rs:18]: > Placing all these interfaces into a single trait is not a good design.
> It's better to use a super trait or an associated type

I agree.



> he suggested that [crate_interface](https://crates.io/crates/crate_interface) can be used here

`crate_interface` is a good solution.

Comment (hky1999) [reply crates/axvmhal/src/lib.rs:18]: > `crate_interface` is a good solution.

But `crate_interface` cannot support super trait or an associated type

**Pull request closed without merging**